mdp/nodes/sfa_nodes.py

```diff
@@ -214,6 +253,137 @@ def get_eta_values(self, t=1):
             self.stop_training()
         return self._refcast(t / (2 * numx.pi) * numx.sqrt(self.d))
 
+    def _rank_deficit_solver_reg(
+            self, A, B = None, eigenvectors=True, turbo="on", range=None,
+            type=1, overwrite=False):
+        """
+        Alternative routine to solve the SFA eigenvalue issue. This can be used
+        in case the normal symeig() call in _stop_training() throws the common
+        SymeigException ('Covariance matrices may be singular').
+
+        This solver applies a moderate regularization to the covariance matrix
+        before applying symeig(). Afterwards it properly detects the rank
+        deficit and filters out malformed features.
+        This procedure is (approximately) as efficient as the ordinary SFA
+        implementation based on plain symeig, because all additional steps are
+        computationally cheap.
+
+        Note:
+        For efficiency reasons it actually modifies the covariance matrix
+        (even if overwrite=False), but the changes are negligible.
+        """
+        if type != 1:
+            raise ValueError('Only type=1 is supported.')
+        # apply some regularization...
+        # The following is equivalent to B += 1e-14*np.eye(B.shape[0]),
+        # but works more in place, i.e. saves memory consumption of np.eye().
+        Bflat = B.reshape(B.shape[0]*B.shape[1])
+        idx = numx.arange(0, len(Bflat), B.shape[0]+1)
+        Bflat[idx] += 1e-12
```

Comment:
I don't think this is a good idea. You are unconditionally adding a value to the whole matrix. The eigenvalues and the eigenvectors are not invariant under translation, so you are effectively changing the solution. Because the value is small in absolute value, you would normally not notice the problem. But SFA is interested in the **smallest** eigenvalues, so you are interfering with exactly those eigenvalues we care most. There is also a problem with numerical stability. If SFA is applied in a polynomial space (say with degree higher than 3), then the eigenvalues will have a dynamic range which is often quite big, with a ratio betwen the biggest and the smallest eigenvalue which sometimes can exceed double numerical precision (1e-13). In my experience these are the cases where the standard eigensolver has problems, because of accumulation of numerical errors. By doing this regularization you are putting the problem under the carpet somehow, but the numers you get have nothing to do with the original problem… I have the feeling that we need a test case which is not fictional, and we should then compare the different approaches.

Reply:
In general I have to say that before going on with this approach we should check that the SVD and the regularization approach are mathematically **and** computationally equivalent to the generalized eigenvalues approach. I am still not completely sure about it, especially in the numerically instable cases, where we care exactly about those eigenvalues that SVD and the regularization are going to discard… That is why until now the preferred and suggested solution to the problem has been to use PCA before SFA: this allows you to control exactly what is begin neglected, and it does that in a space where "throwing away" dimensions is safe.

Reply:
> You are unconditionally adding a value to the whole matrix.

Only to the diagonal, so this procedure enforces full rank.

> The eigenvalues and the eigenvectors are not invariant under translation, so you are effectively changing the solution. Because the value is small in absolute value, you would normally not notice the problem.

I think the regularization of order 1e-12 should not perturb the eigenvalues significantly. However you will get a slightly (i.e. measurably) different eigenvector/extraction matrix. This is actually okay and expectable, because the optimal extraction matrix is not unique if the input data contains linear redundancy. [testSFANode_rank_deficit()](https://github.com/mdp-toolkit/mdp-toolkit/pull/31/files#diff-014aa0eba7c7d99e0021224acf743c50R80) already ensures that the resulting extraction matrix is of equivalent quality.

> But SFA is interested in the smallest eigenvalues, so you are interfering with exactly those eigenvalues we care most.

The implementation distinguishes small eigenvalues introduced by regularization from actual slow features based on the covariance matrix rather than the dcov matrix (see [this line and following](https://github.com/mdp-toolkit/mdp-toolkit/pull/31/files#diff-e0a5be9e597a4c5d554db9950e4155edR285)).
With a covariance matrix B and extraction matrix V the identity `V.T B V == I` [should hold](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.linalg.eigh.html#scipy.linalg.eigh). Actual rank deficit of B (only fixed by regularization) leads to a significant violation of this equation (Dimensions belonging to rank deficit differ from 1 by a threshold ~1e-4, which is really significant).
I believe this is a robust way to filter away purely regularization-induced almost-zero eigenvalues while not significantly changing actual slow-feature-induced ones.

> I have the feeling that we need a test case which is not fictional, and we should then compare the different approaches.

I absolutely agree. Suggestions for a proper test case? Would it be sufficient to use [the data that ordinary SFA is tested on](https://github.com/Stewori/mdp-toolkit/blob/e887365a00efe14307099e159b82231407fc7028/mdp/test/test_SFANode.py#L10) (with some redundancy added)?